File: core/src/main/java/com/cc/core/actions/Actions.java
package com.cc.core.actions;

import com.cc.core.log.KLog;
import com.cc.core.rpc.Rpc;
import com.cc.core.rpc.RpcArgs;
import com.cc.core.utils.StrUtils;
import com.google.gson.annotations.SerializedName;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Actions {
    private static Map<String, Action> actions = new HashMap<>();

    static {
        loadActions();
    }

    private static void loadActions() {
        /*try {
            BufferedReader reader = new BufferedReader(new InputStreamReader(MyApplication.application.getAssets().open("actions.ini")));
            try {
                for (String line = reader.readLine(); line != null; line = reader.readLine()) {
                    String name = line.trim();
                    if (name.length() > 0) addAction(name);
                }
            } finally {
                reader.close();
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }*/
        try {
            for (String a : ActionNames.ACTION_NAMES) {
                addAction(a);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @SuppressWarnings("rawtypes")
    private static void addAction(String className) throws Exception {
        Class action = Class.forName(className);
        if (isAction(action)) {
            put((Action) action.newInstance());
        }
    }

    @SuppressWarnings("rawtypes")
    private static boolean isAction(Class actionCandidate) {
        boolean isImplementation = !actionCandidate.isInterface();
        return isImplementation && Action.class.isAssignableFrom(actionCandidate);
    }

    private static void put(Action action) {
        if (getActions().containsKey(action.key())) {
            Action duplicate = getActions().get(action.key());
            throw new RuntimeException("Found duplicate action key:'" + action.key() + "'. [" + duplicate.getClass().getName() + "," + action.getClass().getName() + "]");
        }
        if (action.key() == null) {
            System.out.println("Skipping " + action.getClass() + ". Key is null.");
        } else {
            getActions().put(action.key(), action);
        }
    }

    public static Action lookup(String key) {
        Action action = getActions().get(key);
        if (action == null) {
            action = new NullAction();
            ((NullAction) action).setMissingKey(key);
        }
        return action;
    }

    public static Action lookup(Class<? extends Action> clazz) {
        for (String key : actions.keySet()) {
            if (actions.get(key).getClass() == clazz) {
                return actions.get(key);
            }
        }

        return new NullAction(clazz.getName());
    }

    public static Map<String, Action> getActions() {
        return actions;
    }

    public static ActionResult execute(Class<? extends Action> clazz, Object... args) {
        Action action = lookup(clazz);
        if (isWechatAction(action)) {
            try {
                return Rpc.call(RpcArgs.newMessage(RawAction.fromAction(action, args)));
            } catch (Exception e) {
                e.printStackTrace();
                return ActionResult.failedResult(e);
            }
        } else {
            return action.execute(args);
        }
    }

    public static String executeCommand(String key, Object... args) {
        Action action = lookup(key);
        KLog.e("--->>>>", "execute action:" + action.key());
        return StrUtils.toJson(action.execute(args));
    }

    public static ActionResult receivedAction(String rawAction) {
        RawAction raw = StrUtils.fromJson(rawAction, RawAction.class);
        KLog.e(raw.toString());
        Action action = lookup(raw.actionName);
        return action.execute(raw.args != null ? raw.args.toArray() : null);
    }

    private static boolean isWechatAction(Action action) {
        return action.key().startsWith("wechat:");
    }

    public static class RawAction {
        @SerializedName("actionName")
        public String actionName;

        @SerializedName("args")
        public List<? extends Object> args;

        @Override
        public String toString() {
            return "RawAction{" +
                    "actionName='" + actionName + '\'' +
                    ", args=" + args +
                    '}';
        }

        public static RawAction fromAction(Action action, Object... args) {
            RawAction a = new RawAction();
            a.actionName = action.key();
            a.args = Arrays.asList(args);
            return a;
        }
    }
}

File: core/src/main/java/com/cc/core/actions/Actions.kt
package com.cc.core.actions

import android.text.TextUtils
import com.cc.core.log.KLog
import com.cc.core.rpc.Rpc
import com.cc.core.rpc.RpcArgs
import com.cc.core.utils.StrUtils
import com.cc.core.wechat.Wechat
import com.cc.core.wechat.Wechat.SupportVersion
import com.cc.core.wechat.Wechat.WechatVersion
import com.google.gson.annotations.SerializedName
import java.util.*

class Actions {
    companion object {

        private val actions : HashMap<String, Action> = HashMap()

        init {
            loadActions()
        }

        private fun loadActions(){
            try {
                for (a in ActionNames.ACTION_NAMES) {
                    addAction(a)
                }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }

        @Throws(Exception::class)
        private fun addAction(className: String) {
            val action = Class.forName(className)
            if (isAction(action)) {
                put(action.newInstance() as Action)
            }
        }

        private fun isAction(actionCandidate: Class<*>): Boolean {
            val isImplementation = !actionCandidate.isInterface
            return isImplementation && Action::class.java.isAssignableFrom(actionCandidate)
        }

        private fun put(action: Action) {
            if (actions.containsKey(action.key())) {
                val duplicate = actions[action.key()]
                throw RuntimeException("Found duplicate action key:'" + action.key() + "'. [" + (duplicate as Action).javaClass.name + "," + action.javaClass.name + "]")
            }

            if (TextUtils.isEmpty(action.key())) {
                KLog.i("Skipping " + action.javaClass.name + ". Key is null.")
            } else {
                actions[action.key().toString()] = action
            }
        }

        fun lookup(key: String?): Action {
            var action: Action? = getActions()[key]
            if (action == null) {
                action = NullAction()
                action.setMissingKey(key)
            }
            return action
        }

        fun lookup(clazz: Class<out Action>): Action {
            actions.entries.forEach {
                if (it.value.javaClass === clazz) {
                    return it.value
                }
            }

            return NullAction(clazz.name)
        }

        fun getActions(): MutableMap<String, Action> {
            return actions
        }

        fun execute(clazz: Class<out Action>, id : String, vararg args: Any?): ActionResult? {
            val action = lookup(clazz)
            return if (isWechatAction(action)) {
                try {
                    Rpc.call(RpcArgs.newMessage(RawAction.fromAction(action, id, *args)))
                } catch (e: Throwable) {
                    e.printStackTrace()
                    ActionResult.failedResult(id, e)
                }

            } else {
                try {
                    action.execute(id, *args)
                } catch (e: Throwable) {
                    e.printStackTrace()
                    ActionResult.failedResult(id, e)
                }
            }
        }

        fun executeCommand(id : String, key: String, vararg args: Any?): String {
            val action = lookup(key)
            return if (isWechatAction(action)) {
                try {
                    StrUtils.toJson(Rpc.call(RpcArgs.newMessage(RawAction.fromAction(action, id, *args))))
                } catch (e: Exception) {
                    e.printStackTrace()
                    StrUtils.toJson(ActionResult.failedResult(id, e))
                }

            } else {
                StrUtils.toJson(action.execute(id, *args))
            }
        }

        fun receivedAction(rawAction: String): ActionResult? {
            val raw = StrUtils.fromJson(rawAction, RawAction::class.java)
            if (!checkSupported()) {
                return ActionResult.failedResult(raw.actionId, "微信版本${WechatVersion}不支持，暂支持版本：${Arrays.toString(SupportVersion)}")
            }
            val action = lookup(raw.actionName)
            return try {
                action.execute(raw.actionId, *raw.args!!.toTypedArray())
            } catch (t : Throwable) {
                KLog.e("Actions", "Failed to execute action:${raw.actionName}, error:${t.localizedMessage}")
                ActionResult.failedResult(raw.actionId, "Failed to execute action:${raw.actionName}, error:${t.localizedMessage}")
            }
        }

        private fun isWechatAction(action: Action): Boolean {
            return StrUtils.stringNotNull(action.key()).startsWith("wechat:")
        }

        private fun checkSupported() : Boolean {
            if (!Arrays.asList(*Wechat.SupportVersion).contains(WechatVersion)) {
                return false
            }
            return true
        }
    }

    class RawAction {
        @SerializedName("id")
        var actionId : String = ""

        @SerializedName("actionName")
        var actionName: String? = null

        @SerializedName("args")
        var args: MutableList<Any?>? = null

        override fun toString(): String {
            return """
                RawAction{actionName='$actionName', args=${args.toString()}
            """
        }

        companion object {

            fun fromAction(action: Action, id: String, vararg args: Any?): RawAction {
                val a = RawAction()
                a.actionName = action.key()
                a.actionId = id

                a.args = mutableListOf()
                for (arg in args) {
                    (a.args as MutableList).add(arg)
                }
                return a
            }
        }
    }
}